refactor: ✨v2 bookings allow non utc start time

shaun-ak opened the pull request:
## What does this PR do?
 
 Users can now send start times in any ISO 8601 format:

- With timezone offset (+05:30, -04:00, etc.)
- With Z (UTC)
- Or just a plain local time (which will be interpreted as the attendee's timezone)


Fixes #21738 (GitHub issue number)
Fixes CAL-5903 (Linear issue number - should be visible at the bottom of the GitHub issue description)

#### Image Demo (if applicable):

With timezone offset (+05:30):
<img width="1145" alt="Screenshot 2025-06-27 at 8 03 42 AM" src="https://github.com/user-attachments/assets/84a1e0fd-cb4c-4f28-9f6e-73a4594955aa" />

Without any timezone, will take from attendee.timezone:
<img width="1145" alt="Screenshot 2025-06-27 at 8 05 10 AM" src="https://github.com/user-attachments/assets/368284a6-43e9-4fea-8a89-ae92d283080c" />

With UTC (Z) at the end:
<img width="1145" alt="Screenshot 2025-06-27 at 8 25 03 AM" src="https://github.com/user-attachments/assets/67f3e026-22e7-4dcd-865b-2573298f65b9" />

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.


    
<!-- This is an auto-generated description by cubic. -->
---

## Summary by cubic
Bookings now accept start times in any ISO 8601 format, including those with time zone offsets, UTC (Z), or local times based on the attendee's time zone.

- **Refactors**
  - Updated booking logic to handle non-UTC start times and correctly interpret time zones.

<!-- End of auto-generated description by cubic. -->

shaun-ak committed:
refactor: v2 bookings allow non utc start time

shaun-ak committed:
refactor: refactor redundant code

shaun-ak committed:
refactor: add hour offset regex

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak commented:
Hey @anikdhabal @kart1ka , just a gentle ping if this could be reviewed.

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

kart1ka reviewed (dismissed):
LGTM

kart1ka committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

anikdhabal committed:
update

supalarry reviewed (changes requested):
Thank you for the contribution - we really appreciate! I think we need to re-think this though:
1. Right now we can submit `attendee.timeZone` and provide an optional time zone within the `start` time - but then we will mix 2 timezones. If you check [input.service.ts](https://github.com/calcom/cal.com/pull/22078/files#diff-0b6b19aec82cfa9d04de462fb7fb6c4397aa56bfa86b13150ea590dbd8cc6f41) line 174 we still pass attedee time zone as the time zone `timeZone: inputBooking.attendee.timeZone,` but if timezone is provided as part of `start` time I think we are mixing things up. That is why I think that
- either user submits timezone as part of `start` OR `attendee.timeZone`. We then need a validation to ensure that one of these is passed and not both at the same time unless timeZone in both are the same. Then we need to make `attendee.timeZone` optional in the input dto [create-booking.input.ts](https://github.com/calcom/cal.com/pull/22078/files#diff-6ab12db3db20850b6c1a2162fde5643f4278fb90b6290e94627005c5698d0af6)
2. Can we add e2e tests in `apps/api/v2/src/ee/bookings/2024-08-13/controllers/e2e/user-bookings.e2e-spec.ts` testing if we can schedule and reschedule correctly if timezone is provided in `start` time instead of `attendee.timeZone`.

Thank you

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
feat: update timezone decision logic

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time

shaun-ak committed:
feat: add timezone guessing from offset

shaun-ak committed:
feat: add test case for scheduling and rescheduling

shaun-ak commented:
Hi @supalarry , thank you for the detailed review.

I have added some more logic as passing "+05:00" etc, i.e offset to the start time needs to be converted to a valid IANA zone, so I have added a function that guesses the timezone if not provided in `attendee.timeZone`. I do not know if this is a valid approach as many valid timezones can have "+05:00" etc as their offsets, the logic always picks up the first element from the guessed zones (this result can have multiple timezones too).

shaun-ak committed:
feat: address ai review

shaun-ak committed:
Merge branch 'main' into v2-bookings-allow-non-utc-start-time
